File: core/src/main/kotlin/com/example/toyracers/car/CarPhysics.kt
package com.example.toyracers.car

import com.example.toyracers.input.PlayerInput
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sin

/** Deterministic arcade car simulation without platform or rendering dependencies. */
class CarPhysics {
    fun update(
        state: CarState,
        config: CarConfig,
        rawInput: PlayerInput,
        deltaSeconds: Float,
    ) {
        require(deltaSeconds >= 0f) { "deltaSeconds must not be negative" }
        if (deltaSeconds == 0f) return

        val input = rawInput.normalized()
        var basis = Basis.fromDegrees(state.rotationDeg)

        // Engine force always follows the car's forward direction.
        state.velocityX += basis.forwardX * config.acceleration * input.throttle * deltaSeconds
        state.velocityY += basis.forwardY * config.acceleration * input.throttle * deltaSeconds

        var longitudinalSpeed = basis.forwardDot(state.velocityX, state.velocityY)
        val lateralSpeedBeforeSteering = basis.rightDot(state.velocityX, state.velocityY)
        if (input.brake > 0f) {
            if (longitudinalSpeed > STOP_EPSILON) {
                longitudinalSpeed =
                    moveToward(
                        longitudinalSpeed,
                        0f,
                        config.brakeForce * input.brake * deltaSeconds,
                    )
            } else {
                longitudinalSpeed -= config.reverseAcceleration * input.brake * deltaSeconds
            }
        }

        longitudinalSpeed =
            moveToward(
                longitudinalSpeed,
                0f,
                config.rollingResistance * deltaSeconds,
            ).coerceIn(-config.maxReverseSpeed, config.maxForwardSpeed)

        val targetDriftAmount =
            targetDriftAmount(
                longitudinalSpeed = longitudinalSpeed,
                lateralSpeed = lateralSpeedBeforeSteering,
                steering = input.steering,
                config = config,
            )
        state.driftAmount =
            moveToward(
                state.driftAmount,
                targetDriftAmount,
                if (targetDriftAmount > state.driftAmount) {
                    config.driftEntryResponse * deltaSeconds
                } else {
                    config.driftExitResponse * deltaSeconds
                },
            )

        val velocityBeforeSteeringX =
            basis.forwardX * longitudinalSpeed + basis.rightX * lateralSpeedBeforeSteering
        val velocityBeforeSteeringY =
            basis.forwardY * longitudinalSpeed + basis.rightY * lateralSpeedBeforeSteering

        // Steering authority rises with speed and reverses naturally while backing up.
        val steeringAuthority = min(abs(longitudinalSpeed) / STEERING_REFERENCE_SPEED, 1f)
        state.angularVelocity = -input.steering *
            config.steeringSpeed *
            interpolate(1f, config.driftSteeringMultiplier, state.driftAmount) *
            steeringAuthority *
            signOrZero(longitudinalSpeed)
        state.rotationDeg = normalizeDegrees(state.rotationDeg + state.angularVelocity * deltaSeconds)

        // Re-project both velocity components after rotating the car, then remove side slip.
        basis = Basis.fromDegrees(state.rotationDeg)
        longitudinalSpeed = basis.forwardDot(velocityBeforeSteeringX, velocityBeforeSteeringY)
        var lateralSpeed = basis.rightDot(velocityBeforeSteeringX, velocityBeforeSteeringY)
        val effectiveGrip =
            config.grip *
                interpolate(
                    1f,
                    config.driftGripMultiplier,
                    state.driftAmount,
                )
        lateralSpeed *= max(0f, 1f - config.lateralFriction * effectiveGrip * deltaSeconds)
        longitudinalSpeed =
            moveToward(
                longitudinalSpeed,
                0f,
                config.driftDrag * state.driftAmount * abs(lateralSpeed) * deltaSeconds,
            )

        longitudinalSpeed = longitudinalSpeed.coerceIn(-config.maxReverseSpeed, config.maxForwardSpeed)
        state.velocityX = basis.forwardX * longitudinalSpeed + basis.rightX * lateralSpeed
        state.velocityY = basis.forwardY * longitudinalSpeed + basis.rightY * lateralSpeed
        state.speed = longitudinalSpeed
        state.lateralSpeed = lateralSpeed
        state.x += state.velocityX * deltaSeconds
        state.y += state.velocityY * deltaSeconds
    }

    private fun targetDriftAmount(
        longitudinalSpeed: Float,
        lateralSpeed: Float,
        steering: Float,
        config: CarConfig,
    ): Float {
        if (
            longitudinalSpeed < config.driftEntrySpeed ||
            steering * lateralSpeed < 0f
        ) {
            return 0f
        }

        val speedRatio =
            (
                (longitudinalSpeed - config.driftEntrySpeed) / config.driftEntrySpeed
            ).coerceIn(0f, 1f)
        val steeringRatio =
            (
                (abs(steering) - config.driftSteeringThreshold) /
                    (1f - config.driftSteeringThreshold)
            ).coerceIn(0f, 1f)
        return speedRatio * steeringRatio
    }

    private fun interpolate(
        from: Float,
        to: Float,
        amount: Float,
    ): Float = from + (to - from) * amount

    private fun moveToward(
        value: Float,
        target: Float,
        amount: Float,
    ): Float =
        when {
            value < target -> min(value + amount, target)
            value > target -> max(value - amount, target)
            else -> target
        }

    private fun signOrZero(value: Float): Float =
        when {
            value > STOP_EPSILON -> 1f
            value < -STOP_EPSILON -> -1f
            else -> 0f
        }

    private fun normalizeDegrees(degrees: Float): Float {
        val wrapped = degrees % 360f
        return if (wrapped < 0f) wrapped + 360f else wrapped
    }

    private data class Basis(
        val forwardX: Float,
        val forwardY: Float,
        val rightX: Float,
        val rightY: Float,
    ) {
        fun forwardDot(
            x: Float,
            y: Float,
        ): Float = x * forwardX + y * forwardY

        fun rightDot(
            x: Float,
            y: Float,
        ): Float = x * rightX + y * rightY

        companion object {
            fun fromDegrees(rotationDeg: Float): Basis {
                val radians = Math.toRadians(rotationDeg.toDouble())
                val forwardX = cos(radians).toFloat()
                val forwardY = sin(radians).toFloat()
                return Basis(
                    forwardX = forwardX,
                    forwardY = forwardY,
                    rightX = -forwardY,
                    rightY = forwardX,
                )
            }
        }
    }

    companion object {
        const val FIXED_DELTA_SECONDS = 1f / 60f
        const val MAX_FRAME_DELTA_SECONDS = 0.25f
        private const val STEERING_REFERENCE_SPEED = 12f
        private const val STOP_EPSILON = 0.01f
    }
}

File: dart/lib/simulation/car/car_physics.dart
import 'dart:math' as math;

import '../input/player_input.dart';
import '../math/float32.dart';
import 'car_config.dart';
import 'car_state.dart';

/// Deterministic reference-compatible arcade vehicle integrator.
///
/// It owns throttle, brake, reverse, steering, grip, drift, and position
/// integration. Callers supply an explicit fixed timestep; no rendering or
/// wall-clock state participates in an update.
final class CarPhysics {
  /// Advances [state] by an explicit fixed [deltaSeconds] using [input].
  void update({
    required CarState state,
    required CarConfig config,
    required PlayerInput input,
    required double deltaSeconds,
  }) {
    final step = Float32.narrow(deltaSeconds);
    if (step < 0) {
      throw ArgumentError.value(
        deltaSeconds,
        'deltaSeconds',
        'must not be negative',
      );
    }
    if (step == 0) {
      return;
    }

    final normalizedInput = input.normalized();
    var basis = _Basis.fromDegrees(state.rotationDegrees);

    state.velocityX = Float32.add(
      state.velocityX,
      Float32.multiply(
        Float32.multiply(
          Float32.multiply(basis.forwardX, config.acceleration),
          normalizedInput.throttle,
        ),
        step,
      ),
    );
    state.velocityY = Float32.add(
      state.velocityY,
      Float32.multiply(
        Float32.multiply(
          Float32.multiply(basis.forwardY, config.acceleration),
          normalizedInput.throttle,
        ),
        step,
      ),
    );

    var longitudinalSpeed = basis.forwardDot(state.velocityX, state.velocityY);
    final lateralSpeedBeforeSteering = basis.rightDot(
      state.velocityX,
      state.velocityY,
    );
    if (normalizedInput.brake > 0) {
      final brakingDistance = Float32.multiply(
        Float32.multiply(config.brakeForce, normalizedInput.brake),
        step,
      );
      if (longitudinalSpeed > _stopEpsilon) {
        longitudinalSpeed = _moveToward(longitudinalSpeed, 0, brakingDistance);
      } else {
        longitudinalSpeed = Float32.subtract(
          longitudinalSpeed,
          Float32.multiply(
            Float32.multiply(config.reverseAcceleration, normalizedInput.brake),
            step,
          ),
        );
      }
    }

    longitudinalSpeed = Float32.clamp(
      _moveToward(
        longitudinalSpeed,
        0,
        Float32.multiply(config.rollingResistance, step),
      ),
      -config.maxReverseSpeed,
      config.maxForwardSpeed,
    );

    final targetDriftAmount = _targetDriftAmount(
      longitudinalSpeed: longitudinalSpeed,
      lateralSpeed: lateralSpeedBeforeSteering,
      steering: normalizedInput.steering,
      config: config,
    );
    state.driftAmount = _moveToward(
      state.driftAmount,
      targetDriftAmount,
      Float32.multiply(
        targetDriftAmount > state.driftAmount
            ? config.driftEntryResponse
            : config.driftExitResponse,
        step,
      ),
    );

    final velocityBeforeSteeringX = Float32.add(
      Float32.multiply(basis.forwardX, longitudinalSpeed),
      Float32.multiply(basis.rightX, lateralSpeedBeforeSteering),
    );
    final velocityBeforeSteeringY = Float32.add(
      Float32.multiply(basis.forwardY, longitudinalSpeed),
      Float32.multiply(basis.rightY, lateralSpeedBeforeSteering),
    );

    final steeringAuthority = Float32.clamp(
      Float32.divide(longitudinalSpeed.abs(), _steeringReferenceSpeed),
      0,
      1,
    );
    state.angularVelocity = Float32.multiply(
      Float32.multiply(
        Float32.multiply(
          Float32.multiply(-normalizedInput.steering, config.steeringSpeed),
          _interpolate(1, config.driftSteeringMultiplier, state.driftAmount),
        ),
        steeringAuthority,
      ),
      _signOrZero(longitudinalSpeed),
    );
    state.rotationDegrees = Float32.wrapDegrees(
      Float32.add(
        state.rotationDegrees,
        Float32.multiply(state.angularVelocity, step),
      ),
    );

    basis = _Basis.fromDegrees(state.rotationDegrees);
    longitudinalSpeed = basis.forwardDot(
      velocityBeforeSteeringX,
      velocityBeforeSteeringY,
    );
    var lateralSpeed = basis.rightDot(
      velocityBeforeSteeringX,
      velocityBeforeSteeringY,
    );
    final effectiveGrip = Float32.multiply(
      config.grip,
      _interpolate(1, config.driftGripMultiplier, state.driftAmount),
    );
    lateralSpeed = Float32.multiply(
      lateralSpeed,
      _nonNegative(
        Float32.subtract(
          1,
          Float32.multiply(
            Float32.multiply(config.lateralFriction, effectiveGrip),
            step,
          ),
        ),
      ),
    );
    longitudinalSpeed = _moveToward(
      longitudinalSpeed,
      0,
      Float32.multiply(
        Float32.multiply(
          Float32.multiply(config.driftDrag, state.driftAmount),
          lateralSpeed.abs(),
        ),
        step,
      ),
    );

    longitudinalSpeed = Float32.clamp(
      longitudinalSpeed,
      -config.maxReverseSpeed,
      config.maxForwardSpeed,
    );
    state.velocityX = Float32.add(
      Float32.multiply(basis.forwardX, longitudinalSpeed),
      Float32.multiply(basis.rightX, lateralSpeed),
    );
    state.velocityY = Float32.add(
      Float32.multiply(basis.forwardY, longitudinalSpeed),
      Float32.multiply(basis.rightY, lateralSpeed),
    );
    state.longitudinalSpeed = longitudinalSpeed;
    state.lateralSpeed = lateralSpeed;
    state.x = Float32.add(state.x, Float32.multiply(state.velocityX, step));
    state.y = Float32.add(state.y, Float32.multiply(state.velocityY, step));
  }

  double _targetDriftAmount({
    required double longitudinalSpeed,
    required double lateralSpeed,
    required double steering,
    required CarConfig config,
  }) {
    if (longitudinalSpeed < config.driftEntrySpeed ||
        Float32.multiply(steering, lateralSpeed) < 0) {
      return 0;
    }

    final speedRatio = Float32.clamp(
      Float32.divide(
        Float32.subtract(longitudinalSpeed, config.driftEntrySpeed),
        config.driftEntrySpeed,
      ),
      0,
      1,
    );
    final steeringRatio = Float32.clamp(
      Float32.divide(
        Float32.subtract(steering.abs(), config.driftSteeringThreshold),
        Float32.subtract(1, config.driftSteeringThreshold),
      ),
      0,
      1,
    );
    return Float32.multiply(speedRatio, steeringRatio);
  }

  static double _interpolate(double from, double to, double amount) =>
      Float32.add(from, Float32.multiply(Float32.subtract(to, from), amount));

  static double _moveToward(double value, double target, double amount) {
    if (value < target) {
      return _minimum(Float32.add(value, amount), target);
    }
    if (value > target) {
      return _maximum(Float32.subtract(value, amount), target);
    }
    return target;
  }

  static double _signOrZero(double value) {
    if (value > _stopEpsilon) {
      return 1;
    }
    if (value < -_stopEpsilon) {
      return -1;
    }
    return 0;
  }

  static double _minimum(double left, double right) =>
      left < right ? left : right;

  static double _maximum(double left, double right) =>
      left > right ? left : right;

  static double _nonNegative(double value) => _maximum(0, value);

  static final double fixedDeltaSeconds = Float32.fixedDeltaSeconds;
  static final double maxFrameDeltaSeconds = Float32.narrow(0.25);
  static final double _steeringReferenceSpeed = Float32.narrow(12);
  static final double _stopEpsilon = Float32.narrow(0.01);
}

final class _Basis {
  const _Basis({
    required this.forwardX,
    required this.forwardY,
    required this.rightX,
    required this.rightY,
  });

  final double forwardX;
  final double forwardY;
  final double rightX;
  final double rightY;

  factory _Basis.fromDegrees(double rotationDegrees) {
    final radians = rotationDegrees * _radiansPerDegree;
    final forwardX = Float32.narrow(math.cos(radians));
    final forwardY = Float32.narrow(math.sin(radians));
    return _Basis(
      forwardX: forwardX,
      forwardY: forwardY,
      rightX: Float32.narrow(-forwardY),
      rightY: forwardX,
    );
  }

  double forwardDot(double x, double y) =>
      Float32.add(Float32.multiply(x, forwardX), Float32.multiply(y, forwardY));

  double rightDot(double x, double y) =>
      Float32.add(Float32.multiply(x, rightX), Float32.multiply(y, rightY));

  static const double _radiansPerDegree = math.pi / 180.0;
}
